### PR TITLE
add error section in report and the rest queries

### LINKED ISSUE #8816: [FEA] Write the rest of the scale test queries
**Is your feature request related to a problem? Please describe.**
This would come after the initial set of queries are done. https://github.com/NVIDIA/spark-rapids/issues/8814 Just so that it can unblock others from being able to run tests and understand what is happening.

It is okay to do the following as multiple different PRs, just because it is a lot of queries. Again I am doing pseudo SQL below. It is not meant to be a copy/paste. Some of these may need to be updated so that the names of the results can be written out to a parquet or ORC file.

q4: Left anti-join lots of ride along columns. 
```
SELECT c_data.* FROM c_data LEFT ANTI JOIN a_facts WHERE primary_a = c_foreign_a
```
q5: Left semi-join lots of ride along columns. 
```
SELECT c_data.* FROM c_data LEFT SEMI JOIN a_facts WHERE primary_a = c_foreign_a
```
q6: Exploding inner large key count equi-join followed by min/max agg. 
```
SELECT c_key2_*, COUNT(1), MIN(c_data_*), MAX(d_data_*) FROM c_data JOIN d_data WHERE c_key2_* = d_key2_* GROUP BY c_key2_*
```
q7: Exploding full outer large key count equi-join followed by min/max agg.
```
SELECT c_key2_*, COUNT(1), MIN(c_data_*), MAX(d_data_*) FROM c_data FULL OUTER JOIN d_data WHERE c_key2_* = d_key2_* GROUP BY c_key2_*
```
q8: Exploding left outer large key count equi-join followed by min/max agg.
```
SELECT c_key2_*, COUNT(1), MIN(c_data_*), MAX(d_data_*) FROM c_data LEFT OUTER JOIN d_data WHERE c_key2_* = d_key2_* GROUP BY c_key2_*
```
q9: Left semi large key count equi-join followed by min/max agg.
```
SELECT c_key2_*, COUNT(1), MIN(c_data_*) FROM c_data LEFT SEMI JOIN d_data WHERE c_key2_* = d_key2_* GROUP BY c_key2_*
```
q10: Left anti large key count equi-join followed by min/max agg.
```
SELECT c_key2_*, COUNT(1), MIN(c_data_*) FROM c_data LEFT ANTI JOIN d_data WHERE c_key2_* = d_key2_* GROUP BY c_key2_*
```
q11: No obvious build side inner equi-join. (Shuffle partitions should be set to `10`)
```
SELECT b_key3_*, e_data_*, b_data_* FROM b_data JOIN e_data WHERE b_key3_* = e_key3_*
```
q12: No obvious build side full outer equi-join. (Shuffle partitions should be set to `10`)
```
SELECT b_key3_*, e_data_*, b_data_* FROM b_data FULL OUTER JOIN e_data WHERE b_key3_* = e_key3_*
```
q13: No obvious build side left outer equi-join. (Shuffle partitions should be set to `10`)
```
SELECT b_key3_*, e_data_*, b_data_* FROM b_data LEFT OUTER JOIN e_data WHERE b_key3_* = e_key3_*
```
q14: both sides large left semi equi-join. (Shuffle partitions should be set to `10`)
```
SELECT b_key3_*, b_data_* FROM b_data LEFT SEMI JOIN e_data WHERE b_key3_* = e_key3_*
```
q15: Both sides large left anti equi-join. (Shuffle partitions should be set to `10`)
```
SELECT b_key3_*, b_data_* FROM b_data LEFT ANTI JOIN e_data WHERE b_key3_* = e_key3_*
```
q16: Extreme skew conditional AST inner join.
```
SELECT a_key4_1, a_data_(1-complexit/2), f_data_(1-complexity/2) FROM a_facts JOIN f_facts WHERE a_key4_1 = f_key4_1 && (a_data_low_unique_1 + f_data_low_unique_1) = 2
```
q17: Extreme skew conditional AST full outer join.
```
SELECT a_key4_1, a_data_(1-complexit/2), f_data_(1-complexit/2) FROM a_fact FULL OUTER JOIN f_fact WHERE a_key4_1 = f_key4_1 && (a_data_low_unique_1 + f_data_low_unique_1) = 2
```
q18: Extreme skew conditional AST left outer join.
```
SELECT a_key4_1, a_data_(1-complexity/2), f_data_(1-complexity/2) FROM a_fact LEFT OUTER JOIN f_fact WHERE a_key4_1 = f_key4_1 && (a_data_low_unique_1 + f_data_low_unique_1) = 2
```
q19: Extreme skew conditional AST left anti join.
```
SELECT a_key4_1, a_data_*, FROM a_fact LEFT ANTI JOIN f_fact WHERE a_key4_1 = f_key4_1 && (a_data_low_unique_1 + f_data_low_unique_1) != 2
```
q20: Extreme skew conditional AST left semi join.
```
SELECT a_key4_1, a_data_* FROM a_fact LEFT SEMI JOIN f_fact WHERE a_key4_1 = f_key4_1 && (a_data_low_unique_1 + f_data_low_unique_1) = 2
```
q21: Extreme skew conditional NON-AST inner join.
```
SELECT a_key4_1, a_data_(1-complexity/2), f_data_(1-complexity/2) FROM a_fact JOIN f_fact WHERE a_key4_1 = f_key4_1 && (length(concat(a_data_low_unique_len_1, f_data_low_unique_len_1))) = 2
```
q22: Group by aggregation, not a lot of combining, but lots of aggregations, and CUDF does sort agg internally.
```
SELECT b_key3_*, complexity number of aggregations that are SUMs of 2 or more numeric data columns multiplied together or MIN/MAX of any data column FROM b_data GROUP BY b_key3_*.
```
q23: Reduction with with lots of aggregations
```
SELECT complexity number of aggregations that are SUMs of 2 or more numeric data columns multiplied together or MIN/MAX of any data column FROM b_data.
```
q24: Group by aggregation with lots of combining, lots of aggs, and CUDF does hash agg internally
```
SELECT g_key3_*, complexity number of aggregations that are SUM/MIN/MAX/AVERAGE/COUNT of 2 or more byte columns cast to int and added, subtracted, multiplied together. FROM g_data GROUP BY g_key3_*
```
q25: collect set group by agg
```
select g_key3_*, collect_set(g_data_enum_1) FROM g_data GROUP BY g_key3_*
```
q26: collect list group by agg with some hope of succeeding.
```
select b_foreign_a, collect_list(b_data_1) FROM b_data GROUP BY b_foreign_a
```
q27: Running Window with skewed partition by columns, and single order by column with small number of basic window ops (min, max, sum, count, average, row_number)
```
select {min(g_data_1), max(g_data_1), sum(g_data_2), count(g_data_3), average(g_data_4), row_number} over (UNBOUNDED PRECEDING TO CURRENT ROW PARTITION BY g_key3_* ORDER BY g_data_row_num_1)
```
q28: Ranged Window with large range (lots of rows preceding and following) skewed partition by columns and single order by column with small number of basic window ops (min, max, sum, count, average)
```
select {min(g_data_1), max(g_data_1), sum(g_data_2), count(g_data_3), average(g_data_4)} over (RANGE BETWEEN 1000 * scale_factor PRECEDING AND 5000 * scale_factor FOLLOWING PARTITION BY g_key3_* ORDER BY g_data_row_num_1)
```
q29: unbounded preceding and following window with skewed partition by columns and single order by column with small number of basic window op (min, max, sum, count, average)
```
select {min(g_data_1), max(g_data_1), sum(g_data_2), count(g_data_3), average(g_data_4)} over (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING PARTITION BY g_key3_* ORDER BY g_data_row_num_1)
```
q30: running window with no partition by columns and single order by column with small number of basic window ops (min, max, sum, count, average, row_number)
```
select {min(g_data_1), max(g_data_1), sum(g_data_2), count(g_data_3), average(g_data_4)} over (ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW ORDER BY g_data_row_num_1)
```
q31: ranged window with large range (lots of rows preceding and following) no partition by columns and single order by column with small number of basic window ops (min, max, sum, count, average)
```
select {min(g_data_1), max(g_data_1), sum(g_data_2), count(g_data_3), average(g_data_4)} over (RANGE BETWEEN 1000 * scale_factor PRECEDING AND 5000 * scale_factor FOLLOWING ORDER BY g_data_row_num_1)
```
q32: unbounded preceding and following window with no partition by columns and single order by column with small number of basic window ops (min, max, sum, count, average)
```
select  {min(g_data_1), max(g_data_1), sum(g_data_2), count(g_data_3), average(g_data_4)}  over (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING ORDER BY g_row_num_1)
```
q33: Lead/Lag window with skewed partition by columns and single order by column.
```
select {lag(g_data_1, 10 * scale_factor, lead(g_data_2, 10 * scale_factor)} OVER (PARTITION BY g_key3_* ORDER BY g_data_row_num_1)
```
q34: Lead/Lag window with no partition by columns and single order by column.
```
select {lag(g_data_1, 10 * scale_factor, lead(g_data_2, 10 * scale_factor)} OVER (ORDER BY g_data_row_num_1)
```
q35: Running window with complexity/2 in partition by columns and complexity/2 in order by columns.
```
select {min(c_data_1), max(c_data_2)} over (ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW PARTITION BY c_key2_(1 to complexity/2) ORDER BY c_key2_(complexity/2 to complexity)
```
q36: unbounded to unbounded window with complexity/2 in partition by columns and complexity/2 in order by columns.
```
select {min(c_data_1), max(c_data_2)} over (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING PARTITION BY c_key2_(1 to complexity/2) ORDER BY c_key2_(complexity/2 to complexity)
```
q37: Running window with simple partition by and order by columns, but complexity window operations as combinations of a few input columns
```
select {complexity aggregations mins/max of any column or SUM of two or more numeric columns multiplied together} over (ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW PARTITION BY c_foreign_a ORDER BY c_data_row_num_1
```
q38: Ranged window with simple partition by and order by columns, but complexity window operations as combinations of a few input columns
```
select {complexity aggregations mins/max of any column or SUM of two or more numeric columns multiplied together} over (RANGE BETWEEN 10 PRECEDING AND 10 FOLLOWING PARTITION BY c_foreign_a ORDER BY c_data_row_num_1
```
q39: unbounded window with simple partition by and order by columns, but complexity window operations as combinations of a few input columns
```
select {complexity aggregations mins/max of any column or SUM of two or more numeric columns multiplied together} over (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING PARTITION BY c_foreign_a ORDER BY c_data_row_num_1
```
q40: COLLECT SET WINDOW (We may never really be able to do this well)
```
select array_sort(collect_set(f_data_low_unique_1)) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW PARTITION BY f_key4_* order by f_data_row_num_1)
```
q41: COLLECT LIST WINDOW (We may never really be able to do this well)
```
select collect_list(f_data_low_unique_1) OVER (ROWS BETWEEN complexity PRECEDING and CURRENT ROW PARTITION BY f_key4_* order by f_data_row_num_1)
```

### LINKED ISSUE #9165: [BUG] Data gen for key groups produces type-mismatch columns
Take `c_data` and `d_data` tables for example.

both of the 2 tables require key group:
- key group 2 columns up to the complexity (each key should show up about 10 times)

and they can be used in query clause like `c_data.c_key2_1 = d_data.d_key2_1`.

In current implementation, the columns that are described as "complexity" number of columns use a random function to pick its column type, and that random is "random" between tables. 

This causes error like 
```
 cannot resolve '(c_data.c_key2_1 = d_data.d_key2_1)' due to data type mismatch: differing types in '(c_data.c_key2_1 = d_data.d_key2_1)' (decimal(7,2) and timestamp).
```

So the fix should be produce a random column type,  but it should apply for both b_data and c_data.

### PR DESCRIPTION
**Pull request by wjxiz1992**
Close #8816 and #9118 
Close #9165 and #9166 
This PR adds in 
- the query content, an error field in the json report file to see what exceptions are thrown during the query run.
Not only the exceptions that a failed query throws, but also the exceptions that a Spark task failed at first but succeeded in retry. (#9118)
- the rest queries for Scale Test.(#8816 )
- a new input option `--dry`: will only print the generated queries and its physical plan(`.exlpain()`)
- add correct candidate columns for key group 2

This PR is still in testing with different scales, but it contains lots of queries, and it requires a lot of review for all of them since they are parsed according to my personal understanding. Put it up for early feedbacks.

put one generated queries for reviewers to help easy check the query correctness or if the queries are expected:
[queries.txt](https://github.com/NVIDIA/spark-rapids/files/12529819/queries.txt)

### CODE REVIEW

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/TaskFailureListener.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.nvidia.spark.rapids.tests.scaletest
```
Comment (andygrove) [inline integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/TaskFailureListener.scala:1]: Missing license header

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/Utils.scala [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.nvidia.spark.rapids.tests.scaletest
```
Comment (andygrove) [inline integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/Utils.scala:1]: Missing license header

**Comment by wjxiz1992** [comment]
Found one blocking bug: https://github.com/NVIDIA/spark-rapids/issues/9165.
And another bug is not blocking but better to fix in this PR as well: https://github.com/NVIDIA/spark-rapids/issues/9166

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/TaskFailureListener.scala (hunk 1 above)
Comment (wjxiz1992) [reply integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/TaskFailureListener.scala:16]: Added.

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/Utils.scala (hunk 2 above)
Comment (wjxiz1992) [reply integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/Utils.scala:17]: Added.

**Comment by wjxiz1992** [comment]
build

**Comment by wjxiz1992** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by wjxiz1992** [comment]
Hi @revans2 I've added all queries into the test suite, please help take a look when you have time. I've also put a query content list `queries.txt` in the attachment in the PR description. 
Let me know if the queries are not generated as expected.

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/QuerySpecs.scala [hunk 3]
```diff
@@ -61,14 +243,14 @@ class QuerySpecs(config: Config, spark: SparkSession) {
     }
   }
 
-  def getCandidateQueries(): Map[String, TestQuery] = {
+  def getCandidateQueries(): mutable.LinkedHashMap[String, TestQuery] = {
```
Comment (revans2) [inline integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/QuerySpecs.scala:246]: why are the maps mutable? Do we plan on changing them after they are constructed?

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/QuerySpecs.scala [hunk 4]
```diff
@@ -77,26 +259,508 @@ class QuerySpecs(config: Config, spark: SparkSession) {
...
+        s"SELECT " +
+          s"${expandKeyGroup3("b_key3")}, " +
+          s"${expandColumnWithRange("e_data", 1, 10)}, " +
+          s"${expandBDataColumns()} " +
+          s"FROM b_data JOIN e_data on " +
+          s"${expandWhereClauseWithAndKeyGroup3("b_key3", "e_key3")}",
+        config.iterations,
+        config.timeout,
+        "No obvious build side inner equi-join. (Shuffle partitions should be set to 10)",
+        shufflePartitions = 10),
```
Comment (revans2) [inline integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/QuerySpecs.scala:398]: Why 10 shuffle partitions? The original proposal was `ceil(scale_factor/50)`. Not that this is wrong, just curious.

**Review by revans2 — APPROVED** [review]
In general it looks good, but I want to spend some time to test it myself. Just to be sure that the scale_factor/etc look correct and that the queries tax the plugin in ways we expect it to.

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/QuerySpecs.scala (hunk 3 above)
Comment (wjxiz1992) [reply integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/QuerySpecs.scala]: I was looking for Map that can reserve insertion order and found LinkedHashMap first. I just found ListMap can also do this, updated.

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/QuerySpecs.scala (hunk 4 above)
Comment (wjxiz1992) [reply integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/scaletest/QuerySpecs.scala]: Oh I was refering to the queries in https://github.com/NVIDIA/spark-rapids/issues/8816. Now updated to what is set in test plan doc.

**Comment by wjxiz1992** [comment]
build

**Comment by wjxiz1992** [comment]
build

**Comment by wjxiz1992** [comment]
> In general it looks good, but I want to spend some time to test it myself. Just to be sure that the scale_factor/etc look correct and that the queries tax the plugin in ways we expect it to.

I can build pipelines to test this, any test preference? According to the test plan doc, I should start to generate `scale_factor=100`, `complexity=300` data. I don't have enough disk space in my local PC so I can do this on our internal cluster.

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
@wjxiz1992 lets check this in and then we can work on figuring out if we need to tweak things. We need to setup a pipeline anyways, so lets get started with that as a separate piece.

**Pull request merged**